src/twisted/application/internet.py

```diff
@@ -523,13 +524,375 @@ def policy(attempt):
+        The service has started connecting.
+        """
+
+    @_machine.state()
+    def _waiting(self):
+        """
+        The service is waiting for the reconnection period
+        before reconnecting.
+        """
+
+    @_machine.state()
+    def _connected(self):
+        """
+        The service is connected.
+        """
+
+    @_machine.state()
+    def _disconnecting(self):
+        """
+        The service is disconnecting after being asked to shutdown.
+        """
+
+    @_machine.state()
+    def _restarting(self):
+        """
+        The service is disconnecting and has been asked to restart.
+        """
+
+    @_machine.state()
+    def _stopped(self):
+        """
+        The service has been stopped an is disconnected.
+        """
+
+    @_machine.input()
+    def start(self):
+        """
+        Start this L{ClientService}, initiating the connection retry loop.
+        """
+
+    @_machine.output()
+    def _connect(self):
+        """
+        Start a connection attempt.
+        """
+        factoryProxy = _DisconnectFactory(self._factory,
+                                          lambda _: self._clientDisconnected())
+
+        self._connectionInProgress = (
+            self._endpoint.connect(factoryProxy)
+            .addCallback(self._connectionMade)
+            .addErrback(lambda _: self._connectionFailed()))
+
+
+    @_machine.output()
+    def _resetFailedAttempts(self):
+        """
+        Reset the number of failed attempts.
+        """
+        self._failedAttempts = 0
+
+
+    @_machine.input()
+    def stop(self):
+        """
+        Stop trying to connect and disconnect any current connection.
+
+        @return: a L{Deferred} that fires when all outstanding connections are
+            closed and all in-progress connection attempts halted.
+        """
+
+    @_machine.output()
+    def _waitForStop(self):
+        """
+        Return a deferred that will fire when the service has
+        finished disconnecting.
+
+        @return: L{Deferred} that fires when the service has finished disconnecting.
+        """
+        self._stopWaiters.append(Deferred())
+        return self._stopWaiters[-1]
+
+
+    @_machine.output()
+    def _stopConnecting(self):
+        """
+        Stop pending connection attempt.
+        """
+        self._connectionInProgress.cancel()
+
+
+    @_machine.output()
+    def _stopRetrying(self):
+        """
+        Stop pending attempt to reconnect.
+        """
+        self._retryCall.cancel()
+        del self._retryCall
+
+
+    @_machine.output()
+    def _disconnect(self):
+        """
+        Disconnect the current connection.
+        """
+        self._currentConnection.transport.loseConnection()
+
+
+    @_machine.input()
+    def _connectionMade(self, protocol):
+        """
+        A connection has been made.
+
+        @param protocol: The protocol of the connection.
+        @type protocol: L{IProtocol}
+        """
+
+    @_machine.output()
```

Comment:
I've found it helpful to group all the "outputs" in a section by themselves, since they are actual methods, as opposed to the input/state declarations, which are empty.  It's easier to read if they're separated a little bit, I think.